README.md
# Yari Fork

Fork of Yari which removes AI help "features".

You can view it on the Web at [https://webdocs.dev](https://webdocs.dev)

## Quickstart

Development on `yari` involves updating the machinery that renders MDN content
or improving the structure and styling of the MDN UI (e.g. the styling of the
header). If you are more interested in contributing to the MDN content, you
should check out the [content](https://github.com/mdn/content) repo README
instead.

Before you can start working with Yari, you need to:

<!-- Peterbe, Feb 2021: There appears to be a bug in Prettier for .md files
    that forces in a second (extra) whitespace after the `1.` here.
    That breaks `markdownlint` *and* `prettier --check`. Annoying.
    So for now let's make an exception. -->
<!-- markdownlint-disable list-marker-space -->

1.  Install [git](https://git-scm.com/), [Node.js](https://nodejs.org), and
    [Yarn 1](https://classic.yarnpkg.com/en/docs/install). Alternatively, if you
    have [Nix](https://nixos.org) installed on your system, you can use the dev
    shell provided in the `flake.nix` file at the root of this repo with
    `nix develop`, to quickly set up an environment with all of the required
    tools. If you haven't set up Nix to use flakes yet, check the
    [NixOS Wiki](https://nixos.wiki/wiki/Flakes#Enable_flakes) for instructions.

1.  Clone this repository, as well as the MDN
    [content](https://github.com/mdn/content) repository and the MDN
    [translated-content](https://github.com/mdn/translated-content) repository
    if translations are desired:

        git clone https://github.com/webdocs-dev/yari
        git clone https://github.com/mdn/content
        # git clone https://github.com/mdn/translated-content

<!-- markdownlint-enable list-marker-space -->

To run Yari locally, you'll first need to install its dependencies and build the
app locally. Do this like so:

    cd yari
    cp -r libs cloud-function/src/internal # (not in mdn/yari but seems needed?)
    yarn install

Now copy the `.env-dist` file to `.env`:

    cp .env-dist .env

If you followed the instructions above and cloned the `content` repo as a
sibling of your `yari` repo, the `CONTENT_ROOT` environment variable is already
set and Yari will be able to find the content it needs to render.

Also, if you want translated pages, be sure to uncomment the
`TRANSLATED_CONTENT_ROOT` variable.

At this point, you can get started. Run the following lines to compile required
files, start the Yari web server running, and open it in your browser:

    yarn dev
    open http://localhost:3000

If you prefer you can use `yarn start`, which will re-use any previously
compiled files; this is "riskier" but faster. `yarn dev` always ensures that
everything is up-to-date.

The `yarn start` command also starts a server with slightly different behavior —
it doesn't automatically reload when its source code files change, so use with
caution.

See also our [reviewing guide](docs/REVIEWING.md) for information on how to
review Yari changes.

### Pull request requirements

Firstly, thank you for your interest in contributing to Yari! We do have a few
requirements when it comes to pull requests:

1. Please make use of a
   [feature branch workflow](https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow).
2. We prefer if you use the
   [conventional commits format](https://www.conventionalcommits.org/) when
   making pull requests.
3. Lastly, we require that all commits are signed. Please see the documentation
   [about signed commits](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification)
   and
   [how to sign yours](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits)
   on GitHub.

Thank you for your understanding! We look forward to your contributions.

### How to stay up-to-date

Periodically, the code and the content changes. Make sure you stay up-to-date
with something along the following lines (replace `yari-origin` with whatever
you called [the remote location](https://git-scm.com/docs/git-remote) of the
original yari repo):

    git pull yari-origin main
    yarn
    yarn dev

When you embark on making a change, do it on a new branch, for example
`git checkout -b my-new-branch`.

## Hosting the docs as a static site

Make sure you have [Rust](https://rust-lang.org) installed. Then after copying
`.env-dist` to `.env`, run

    yarn build:dist
    yarn build:prepare
    yarn build:static # (this will take a long time)

If the build fails with an error message about running out of memory, try
setting the environment variable `NODE_OPTIONS='--max-old-space-size=4096'`
(replacing 4096 with a "safe" maximum memory usage in megabytes).

Then you can just host the directory `client/build` using a static server. Make
sure you set the 404 page to `en-us/_spas/404.html` (this file includes a hack
to convert URLs with uppercase letters to lowercase so that e.g.
`/en-US/docs/Web` works).

For example, you can host the web docs on port 5042 using Apache2 as follows:

    # (as root from the directory of yari)
    rm -rf /var/www/web-docs
    mkdir -p /var/www
    cp -r client/build /var/www/web-docs
    cat <<EOF > /etc/apache2/sites-available/web-docs.conf
    <VirtualHost *:5042>
        DocumentRoot /var/www/web-docs
        ErrorDocument 404 /en-us/_spas/404.html
    </VirtualHost>
    <Directory /var/www/web-docs>
        RemoveHandler .var
    </Directory>
    EOF
    ln -sf ../sites-available/web-docs.conf /etc/apache2/sites-enabled/web-docs.conf

Then add `Listen 5042` next to `Listen 80` in `/etc/apache2/ports.conf` and run
`systemctl restart apache2`. Now you can visit `http://localhost:5042` to view
the web docs offline.

## License

All source code is [MPL-2.0](https://www.mozilla.org/en-US/MPL/2.0/).

For content, see its
[license](https://github.com/mdn/content/blob/main/LICENSE.md) in the
[mdn/content repository](https://github.com/mdn/content).

## Supported Platforms

`yari` runs on Linux in CI, and when building for Production.

We also support Windows and MacOS, however we don't aim to proactively catch
issues with CI on those platforms. If bugs arise, we welcome issues being filed,
or PRs being opened to fix them.

## How it works

Yari does a number of things, the most important of which is to render and serve
the MDN content found in the [content repo](https://github.com/mdn/content).
Each document is stored as an `index.md` (recommended) or `index.html` file that
contains metadata presented as YAML
[front-matter](https://github.com/mdn/content#fundamental-concepts) followed by
the document source.

The builder converts these "source files" into "build files" using a CLI tool
that iterates over the files, builds the HTML, and lastly packages it up with
the front-end code, ready to be served as static files.

## Development

The `yarn start` command encapsulates the front-end dev server (on
<http://localhost:3000>) and the `server` (on <http://localhost:5042>).

All the sub-commands of `yarn start` can be broken down and run individually if
you want to work more rapidly.

### Setting up `$EDITOR`

If you configure an environment variable called `EDITOR`, either on your system
as a whole or in the root `.env` file, it can be used in the development server
to link to sources which, when clicked, open in your preferred editor/IDE. For
example, in the root of the repo you could run:

    echo 'EDITOR=code' >> .env

Now clicking certain links will open files directly in the currently open VS
Code IDE (replace `code` in the above command with a different text editor name
if needed, e.g. `atom` or whatever). To test it, view any document on
<http://localhost:3000> and click the "Open in your editor" button.

### How the server works

The `server` has two main jobs:

1. Simulate serving the site (e.g. from a server, S3 or a CDN).
2. Trigger builds of documents that haven't been built, by URL.

### Linting

All JavaScript and TypeScript code needs to be formatted with `prettier` and
it's easy to test this with:

    yarn prettier-check

And conveniently, if you're not even interested in what the flaws were, run:

    yarn prettier-format

When you ran `yarn` for the first time (`yarn` is an alias for `yarn install`)
it automatically sets up a `git` pre-commit hook that uses `lint-staged` — a
wrapper for `prettier` that checks only the files in the git commit.

If you have doubts about formatting, submit your pull request anyway. If you
have formatting flaws, the
[pull request checks](https://github.com/features/actions) should catch it.

### Upgrading Packages

We maintain the dependencies using `Dependabot` in GitHub but if you want to
manually upgrade them you can use:

    yarn upgrade-interactive --latest

### Sharing your dev environment with `ngrok`

[`ngrok`](https://ngrok.com/) allows you to start an HTTP proxy server from the
web into your Yari server. This can be useful for testing your current build
using external tools like BrowserStack, WebPageTest, or Google Translate, or to
simply show a friend what you're up to. Obviously it'll never be faster than
your uplink Internet connection but it should be fairly feature-complete.

1. [Create in account on Ngrok.com](https://dashboard.ngrok.com/signup)
2. [Download the executable](https://ngrok.com/download)
3. Start your Yari server with `yarn start` in one terminal
4. Start the `ngrok` executable with: `/path/to/your/ngrok http 5042`

This will display something like this:

    Session Status                online
    Account                       (Plan: Free)
    Version                       2.3.35
    Region                        United States (us)
    Web Interface                 http://127.0.0.1:4040
    Forwarding                    http://920ba2108da8.ngrok.io -> http://localhost:5042
    Forwarding                    https://920ba2108da8.ngrok.io -> http://localhost:5042

    Connections                   ttl     opn     rt1     rt5     p50     p90
                                  0       0       0.00    0.00    0.00    0.00

Now, take that "Forwarding" URL (`https://920ba2108da8.ngrok.io` in this
example) and share it.

## Building

The `server` builds content automatically (on-the-fly) when you're viewing
pages, but you can pre-emptively build all the content in advance if desired.
One potential advantage is that you can get a more complete list of all possible
"flaws" across all documents before you even visit them.

The most fundamental CLI command is:

    yarn build

### What gets built

Every `index.html` becomes two files:

- `index.html` — a fully formed and complete HTML file
- `index.json` — the state information React needs to build the page in the
  client

### Flaw checks

When building you can enable specific "flaw checks" and their level of handling.
Some flaws are "cosmetic" and some are more severe but they should never block a
full build.

More information about how to set flaws can be found in `docs/envvars.md`.

Essentially, the default is to _warn_ about any flaw and you can see those flaws
when using <http://localhost:3000>. For completed builds, all flaws are ignored.
This makes the build faster and there's also no good place to display the flaws
in a production-grade build.

**In the future**, we might make the default flaw level `error` instead. That
means that any new edits to (or creation of) any document will break in
continuous integration if there's a single flaw and the onus will be on you to
fix it.

## Icons and logos

The favicon is generated from `client/src/assets/m-icon.svg` or
`client/src/assets/webdocs-dev-logo.svg` according to `REACT_APP_ORGANIZATION`.
These are automatically converted to png and ico files by `yarn build:prepare`.

## Contact

If you want to talk to us, ask questions, and find out more, join the discussion
on the
[MDN Web Docs chat room](https://chat.mozilla.org/#/room/#mdn:mozilla.org) on
[Matrix](https://wiki.mozilla.org/Matrix).

## Troubleshooting

Some common issues and how to resolve them.

### `Error: ENOSPC: System limit for number of file watchers reached`

There are two options to resolve this.

1. Disable the watcher via
   [`REACT_APP_NO_WATCHER`](docs/envvars.md#react_app_no_watcher)

   `echo REACT_APP_NO_WATCHER=true >> .env`

2. Increase `max_user_watches`:\
   See <https://github.com/guard/listen#increasing-the-amount-of-inotify-watchers>

### `Error: Cannot find module 'levenary'`

We can't know for sure what's causing this error but speculate a bug in how
`yarn` fails to resolve if certain `@babel` helper libs should install its own
sub-dependencies. A sure way to solve it is to run:

    rm -fr node_modules
    yarn install

### `Error: listen EADDRINUSE: address already in use :::5042`

The default server port `:5042` might be in use by another process. To resolve
this, you can pick any unused port (e.g., 6000) and run the following:

    echo SERVER_PORT=6000 >> .env


CHANGELOG.md
# Changelog

## [2.28.3](https://github.com/mdn/yari/compare/v2.28.2...v2.28.3) (2023-07-18)


### Bug Fixes

* **build:** copy fallback images from en-US to l10n ([#7917](https://github.com/mdn/yari/issues/7917)) ([86a4e6f](https://github.com/mdn/yari/commit/86a4e6fbaaa1ad85a8c47c065f4c15eb4e6d4efc))
* **cloud-function:** remove deprecated X-XSS-Protection header ([#9240](https://github.com/mdn/yari/issues/9240)) ([4a02f2e](https://github.com/mdn/yari/commit/4a02f2ed871fbef498f16862dc0adc238bfae58c))
* **deps-dev:** add @babel/plugin-proposal-private-property-in-object ([#9338](https://github.com/mdn/yari/issues/9338)) ([0b4f31a](https://github.com/mdn/yari/commit/0b4f31ac84bab7cd3048467e20c3643e486c385d))
* **flaws:** stop reporting absolute MDN blog links as broken ([#9080](https://github.com/mdn/yari/issues/9080)) ([e59323f](https://github.com/mdn/yari/commit/e59323f9409ec6294d6addbbc52f062a62b167d1))


### Enhancements

* **release-please:** include chore/enhance commits ([#9339](https://github.com/mdn/yari/issues/9339)) ([ec97a3d](https://github.com/mdn/yari/commit/ec97a3d837660a0d228db724cb162ac6aed89d62))


### Miscellaneous

* **deps-dev:** bump @playwright/test from 1.36.0 to 1.36.1 ([#9315](https://github.com/mdn/yari/issues/9315)) ([df09d78](https://github.com/mdn/yari/commit/df09d78d6b72246fdb849d9cc65aafb37fb9716f))
* **deps-dev:** bump @swc/core from 1.3.69 to 1.3.70 ([#9329](https://github.com/mdn/yari/issues/9329)) ([8f5348b](https://github.com/mdn/yari/commit/8f5348be763fde399a4d98ea3495fe57d20956f7))
* **deps-dev:** bump @types/mdast from 3.0.11 to 4.0.0 ([#9251](https://github.com/mdn/yari/issues/9251)) ([a3406e5](https://github.com/mdn/yari/commit/a3406e584003b593cc5ec2fde4d6a02dc7d5e73f))
* **deps-dev:** bump black from 23.3.0 to 23.7.0 in /deployer ([#9312](https://github.com/mdn/yari/issues/9312)) ([34db2f5](https://github.com/mdn/yari/commit/34db2f5bb74a4f92d0f1ceecae7c03edc6db1c5d))
* **deps-dev:** bump black in /testing/integration ([#9310](https://github.com/mdn/yari/issues/9310)) ([5e45251](https://github.com/mdn/yari/commit/5e45251370e7bb1dccf42ff0d72840740001edbc))
* **deps-dev:** bump eslint from 8.44.0 to 8.45.0 ([#9316](https://github.com/mdn/yari/issues/9316)) ([f67677d](https://github.com/mdn/yari/commit/f67677d1ee27c255665437fa2e61f78c456f7831))
* **deps-dev:** bump eslint-plugin-unicorn from 47.0.0 to 48.0.0 ([#9317](https://github.com/mdn/yari/issues/9317)) ([8661a5c](https://github.com/mdn/yari/commit/8661a5c5701a322b4b04e524fa5bbfa65738c07d))
* **deps-dev:** bump html-validate from 7.18.0 to 8.0.5 ([#9089](https://github.com/mdn/yari/issues/9089)) ([c77a3eb](https://github.com/mdn/yari/commit/c77a3ebef8f1382ca2769c3ed3981bb069d05900))
* **deps-dev:** bump postcss-preset-env from 8.5.1 to 9.0.0 ([#9217](https://github.com/mdn/yari/issues/9217)) ([d8b361c](https://github.com/mdn/yari/commit/d8b361c67663ee06f28c5db270d5b5863091bf2e))
* **deps-dev:** bump react-router-dom from 6.14.1 to 6.14.2 ([#9328](https://github.com/mdn/yari/issues/9328)) ([60974a2](https://github.com/mdn/yari/commit/60974a25bf5db00aa770afa8176d0c1304f568be))
* **deps-dev:** bump stylelint-config-recommended from 12.0.0 to 13.0.0 ([#9247](https://github.com/mdn/yari/issues/9247)) ([e6487bc](https://github.com/mdn/yari/commit/e6487bc297d891bffe3721cf328314f02eb3d7c8))
* **deps-dev:** bump typescript from 5.0.4 to 5.1.6 ([#9196](https://github.com/mdn/yari/issues/9196)) ([c41afe5](https://github.com/mdn/yari/commit/c41afe5585f91be7d0d3953368d242e668177c4c))
* **deps-dev:** bump webpack from 5.88.1 to 5.88.2 ([#9333](https://github.com/mdn/yari/issues/9333)) ([5a4ca67](https://github.com/mdn/yari/commit/5a4ca676ba22fe7361e55f3c7d417ae7decdf349))
* **deps-dev:** bump webpack from 5.88.1 to 5.88.2 in /client/pwa ([#9334](https://github.com/mdn/yari/issues/9334)) ([369f103](https://github.com/mdn/yari/commit/369f103eb3ef7df35821d1d5914206162fa7d77a))
* **deps-dev:** bump workbox-webpack-plugin from 6.6.1 to 7.0.0 ([#8990](https://github.com/mdn/yari/issues/8990)) ([b443c6d](https://github.com/mdn/yari/commit/b443c6de256d6d880391840654b756616e175144))
* **deps-dev:** remove @types/hast ([#9335](https://github.com/mdn/yari/issues/9335)) ([933dfc3](https://github.com/mdn/yari/commit/933dfc383ebffdade99e75f51650ace2d1324c05))
* **deps:** bump aws-actions/configure-aws-credentials ([#9091](https://github.com/mdn/yari/issues/9091)) ([5224452](https://github.com/mdn/yari/commit/52244529e364cc8a6d9d2fa23b4a290e12165013))
* **deps:** bump boto3 from 1.28.1 to 1.28.3 in /deployer ([#9311](https://github.com/mdn/yari/issues/9311)) ([914b9ef](https://github.com/mdn/yari/commit/914b9ef4dbd634135a1b91cdba74f6d99a425ebc))
* **deps:** bump click from 8.1.4 to 8.1.5 in /deployer ([#9314](https://github.com/mdn/yari/issues/9314)) ([f1f7cb3](https://github.com/mdn/yari/commit/f1f7cb342331edfdf311f8b25ce7a34cbf5bb55e))
* **deps:** bump cryptography from 39.0.2 to 41.0.0 in /deployer ([#9006](https://github.com/mdn/yari/issues/9006)) ([db086f5](https://github.com/mdn/yari/commit/db086f502d9254dfe390432729e249cdf8e59dc5))
* **deps:** bump got from 12.6.0 to 13.0.0 ([#8959](https://github.com/mdn/yari/issues/8959)) ([d1511c5](https://github.com/mdn/yari/commit/d1511c524a851aa05547f7c4a24497941b019846))
* **deps:** bump inquirer from 9.2.7 to 9.2.8 ([#9297](https://github.com/mdn/yari/issues/9297)) ([102a212](https://github.com/mdn/yari/commit/102a212ab08fb2987b0e917541a6ec71c7aa7d58))
* **deps:** bump lru-cache from 9.1.2 to 10.0.0 ([#9104](https://github.com/mdn/yari/issues/9104)) ([aac219a](https://github.com/mdn/yari/commit/aac219a4c86357900ae7ef292aa5ea73efc1eeb6))
* **deps:** bump mdast-util-from-markdown from 1.3.1 to 2.0.0 ([#9266](https://github.com/mdn/yari/issues/9266)) ([f2c7a3b](https://github.com/mdn/yari/commit/f2c7a3b6ff52b410a95b8730c271b75b18d90489))

server/index.ts
#!/usr/bin/env node
import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";

import chalk from "chalk";
import express from "express";
import send from "send";
import { createProxyMiddleware } from "http-proxy-middleware";
import cookieParser from "cookie-parser";
import openEditor from "open-editor";
import { getBCDDataForPath } from "@mdn/bcd-utils-api";
import sanitizeFilename from "sanitize-filename";

import {
  buildDocument,
  buildLiveSamplePageFromURL,
  renderContributorsTxt,
} from "../build/index.js";
import { findTranslations } from "../content/translations.js";
import { Document, Redirect, FileAttachment } from "../content/index.js";
import {
  ANY_ATTACHMENT_REGEXP,
  CSP_VALUE,
  DEFAULT_LOCALE,
  PLAYGROUND_UNSAFE_CSP_VALUE,
} from "../libs/constants/index.js";
import {
  STATIC_ROOT,
  PROXY_HOSTNAME,
  FAKE_V1_API,
  CONTENT_HOSTNAME,
  OFFLINE_CONTENT,
  CONTENT_ROOT,
  CONTENT_TRANSLATED_ROOT,
  BLOG_ROOT,
  BLOG_IS_ENABLED,
} from "../libs/env/index.js";

import documentRouter from "./document.js";
import fakeV1APIRouter from "./fake-v1-api.js";
import { searchIndexRoute } from "./search-index.js";
import flawsRoute from "./flaws.js";
import { router as translationsRouter } from "./translations.js";
import { staticMiddlewares, originRequestMiddleware } from "./middlewares.js";
import { MEMOIZE_INVALIDATE, getRoot } from "../content/utils.js";

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import { renderHTML } from "../ssr/dist/main.js";
import {
  allPostFrontmatter,
  findPostLiveSampleBySlug,
  findPostBySlug,
  findPostPathBySlug,
} from "../build/blog.js";

async function buildDocumentFromURL(url: string) {
  const document = Document.findByURL(url);
  if (!document) {
    return null;
  }
  const documentOptions = {};
  if (CONTENT_TRANSLATED_ROOT) {
    // When you're running the dev server and build documents
    // every time a URL is requested, you won't have had the chance to do
    // the phase that happens when you do a regular `yarn build`.
    document.translations = findTranslations(
      document.metadata.slug,
      document.metadata.locale
    );
  }
  return await buildDocument(document, documentOptions);
}

const app = express();

const bcdRouter = express.Router({ caseSensitive: true });

// Note that this route will only get hit if .env has this: REACT_APP_BCD_BASE_URL=""
bcdRouter.get("/api/v0/current/:path.json", async (req, res) => {
  const data = getBCDDataForPath(req.params.path);
  return data ? res.json(data) : res.status(404).send("BCD path not found");
});

bcdRouter.use(
  "/updates/v0/",
  createProxyMiddleware({
    target: "http://localhost:8080",
    pathRewrite: (path) => path.replace("/bcd/updates/v0/", "/"),
  })
);

app.use("/bcd", bcdRouter);

// Depending on if FAKE_V1_API is set, we either respond with JSON based
// on `.json` files on disk or we proxy the requests to Kuma.
const target = `${
  ["developer.mozilla.org", "developer.allizom.org"].includes(PROXY_HOSTNAME)
    ? "https://"
    : "http://"
}${PROXY_HOSTNAME}`;
const proxy = FAKE_V1_API
  ? fakeV1APIRouter
  : createProxyMiddleware({
      target,
      changeOrigin: true,
      // proxyTimeout: 20000,
      // timeout: 20000,
    });

const stageApiProxy = createProxyMiddleware({
  target: `https://developer.allizom.org`,
  changeOrigin: true,
  proxyTimeout: 20000,
  timeout: 20000,
  headers: {
    Connection: "keep-alive",
  },
});

const contentProxy =
  CONTENT_HOSTNAME &&
  createProxyMiddleware({
    target: `https://${CONTENT_HOSTNAME}`,
    changeOrigin: true,
    // proxyTimeout: 20000,
    // timeout: 20000,
  });

app.use("/pong/*", stageApiProxy);
app.use("/pimg/*", stageApiProxy);
app.use("/api/v1/stripe/plans", stageApiProxy);
app.use("/api/*", proxy);
// This is an exception and it's only ever relevant in development.
app.use("/users/*", proxy);

// The proxy middleware has to come before all other middleware to avoid modifying the requests we proxy.

app.use(express.json());

// Needed because we read cookies in the code that mimics what we do in Lambda@Edge.
app.use(cookieParser());

app.use(originRequestMiddleware);

app.use(staticMiddlewares);

app.use(express.urlencoded({ extended: true }));

app.post(
  "/csp-violation-capture",
  express.json({ type: "application/csp-report" }),
  (req, res) => {
    const report = req.body["csp-report"];
    console.warn(
      chalk.yellow(
        "CSP violation for directive",
        report["violated-directive"],
        "which blocked:",
        report["blocked-uri"]
      )
    );
    res.sendStatus(200);
  }
);

app.use("/_document", documentRouter);

app.get("/_open", (req, res) => {
  const { line, column, filepath, url } = req.query as {
    line: string;
    column: string;
    filepath: string;
    url: string;
  };
  // Sometimes that 'filepath' query string parameter is a full absolute
  // filepath (e.g. /Users/peterbe/yari/content.../index.html), which usually
  // happens when you this is used from the displayed flaws on a preview
  // page.
  // But sometimes, it's a relative path and if so, it's always relative
  // to the main builder source.
  let absoluteFilepath;
  if (filepath) {
    if (fs.existsSync(filepath)) {
      absoluteFilepath = filepath;
    } else {
      const [locale] = filepath.split(path.sep);
      const root = getRoot(locale);
      absoluteFilepath = path.join(root, filepath);
    }
  } else if (url) {
    const document = Document.findByURL(url);
    if (!document) {
      res.status(410).send(`No known document by the URL '${url}'\n`);
    }
    absoluteFilepath = document.fileInfo.path;
  } else {
    throw new Error("No .filepath or .url in the request query");
  }

  // Double-check that the file can be found.
  if (!fs.existsSync(absoluteFilepath)) {
    return res.status(400).send(`${absoluteFilepath} does not exist on disk.`);
  }

  let spec = absoluteFilepath;
  if (line) {
    spec += `:${parseInt(line)}`;
    if (column) {
      spec += `:${parseInt(column)}`;
    }
  }
  openEditor([spec]);
  res.status(200).send(`Tried to open ${spec} in ${process.env.EDITOR}`);
});

app.use("/:locale/search-index.json", searchIndexRoute);

app.get("/_flaws", flawsRoute);

app.use("/_translations", translationsRouter);

app.get("/*/contributors.txt", async (req, res) => {
  const url = req.path.replace(/\/contributors\.txt$/, "");
  const document = Document.findByURL(url);
  res.setHeader("content-type", "text/plain");
  if (!document) {
    return res.status(404).send(`Document not found by URL (${url})`);
  }
  const { doc: builtDocument } = await buildDocument(document);
  res.send(
    renderContributorsTxt(
      document.metadata.contributors,
      builtDocument.source.github_url.replace("/blob/", "/commits/")
    )
  );
});

if (BLOG_IS_ENABLED) {
  app.get("/:locale/blog/index.json", async (_, res) => {
    const posts = await allPostFrontmatter(
      { includeUnpublished: true },
      MEMOIZE_INVALIDATE
    );
    return res.json({ hyData: { posts } });
  });
  app.get("/:locale/blog/author/:slug/:asset", async (req, res) => {
    const { slug, asset } = req.params;
    return send(
      req,
      path.resolve(
        BLOG_ROOT,
        "..",
        "authors",
        sanitizeFilename(slug),
        sanitizeFilename(asset)
      )
    ).pipe(res);
  });
  app.get("/:locale/blog/:slug/index.json", async (req, res) => {
    const { slug } = req.params;
    const data = await findPostBySlug(slug);
    if (!data) {
      return res.status(404).send("Nothing here 🤷‍♂️");
    }
    return res.json(data);
  });
  app.get(
    ["/:locale/blog/:slug/runner.html", "/:locale/blog/:slug/runner.html"],
    async (req, res) => {
      return res
        .setHeader("Content-Security-Policy", PLAYGROUND_UNSAFE_CSP_VALUE)
        .status(200)
        .sendFile(path.join(STATIC_ROOT, "runner.html"));
    }
  );
  app.get("/:locale/blog/:slug/_sample_.:id.html", async (req, res) => {
    const { slug, id } = req.params;
    try {
      return res.send(await findPostLiveSampleBySlug(slug, id));
    } catch (e) {
      return res.status(404).send(e.toString());
    }
  });
  app.get("/:locale/blog/:slug/:asset", async (req, res) => {
    const { slug, asset } = req.params;
    const p = findPostPathBySlug(slug);
    if (p) {
      return send(
        req,
        path.resolve(path.join(p, sanitizeFilename(asset)))
      ).pipe(res);
    }
    return res.status(404).send("Nothing here 🤷‍♂️");
  });
}

app.get("/*", async (req, res, ...args) => {
  const parsedUrl = new URL(req.url, `http://localhost:${PORT}`);
  if (req.url.startsWith("/_")) {
    // URLs starting with _ is exclusively for the meta-work and if there
    // isn't already a handler, it's something wrong.
    return res.status(404).send("Page not found");
  }

  // If the catch-all gets one of these something's gone wrong
  if (req.url.startsWith("/static")) {
    return res.status(404).send("Page not found");
  }
  if (OFFLINE_CONTENT) {
    return res.status(404).send("Offline");
  }
  if (contentProxy) {
    console.log(`proxying: ${req.url}`);
    return contentProxy(req, res, ...args);
  }

  if (parsedUrl.pathname.endsWith("/runner.html")) {
    return res
      .setHeader("Content-Security-Policy", PLAYGROUND_UNSAFE_CSP_VALUE)
      .status(200)
      .sendFile(path.join(STATIC_ROOT, "runner.html"));
  }
  if (req.url.includes("/_sample_.")) {
    try {
      return res
        .setHeader("Content-Security-Policy", PLAYGROUND_UNSAFE_CSP_VALUE)
        .status(200)
        .send(await buildLiveSamplePageFromURL(req.path));
    } catch (e) {
      return res.status(404).send(e.toString());
    }
  }

  if (!req.url.includes("/docs/")) {
    // If it's a known SPA, like `/en-US/search` then that should have been
    // matched to its file and not end up here in the catchall handler.
    // Simulate what we do in the Lambda@Edge.
    return res
      .status(404)
      .sendFile(path.join(STATIC_ROOT, "en-us", "_spas", "404.html"));
  }

  if (ANY_ATTACHMENT_REGEXP.test(req.path)) {
    // Remember, FileAttachment.findByURLWithFallback() will return the absolute file path
    // iff it exists on disk.
    // Using a "fallback" strategy here so that images embedded in live samples
    // are resolved if they exist in en-US but not in <locale>
    const filePath = FileAttachment.findByURLWithFallback(req.path);
    if (filePath) {
      // The second parameter to `send()` has to be either a full absolute
      // path or a path that doesn't start with `../` otherwise you'd
      // get a 403 Forbidden.
      // See https://github.com/mdn/yari/issues/1297
      return send(req, path.resolve(filePath)).pipe(res);
    }
    return res.status(404).send("File not found on disk");
  }

  let lookupURL = decodeURI(req.path);
  let extraSuffix = "";
  let isMetadata = false;
  let isDocument = false;

  if (req.path.endsWith("index.json")) {
    // It's a bit special then.
    // The URL like me something like
    // /en-US/docs/HTML/Global_attributes/index.json
    // and that won't be found in getRedirectUrl() since that doesn't
    // index things with the '/index.json' suffix. So we need to
    // temporarily remove it and remember to but it back when we're done.
    isDocument = true;
    extraSuffix = "/index.json";
    lookupURL = lookupURL.replace(extraSuffix, "");
  } else if (req.path.endsWith("metadata.json")) {
    isMetadata = true;
    extraSuffix = "/metadata.json";
    lookupURL = lookupURL.replace(extraSuffix, "");
  }

  const isJSONRequest = extraSuffix.endsWith(".json");

  let document;
  try {
    console.time(`buildDocumentFromURL(${lookupURL})`);
    const built = await buildDocumentFromURL(lookupURL);
    if (built) {
      document = built.doc;
    } else if (
      lookupURL.split("/")[1] &&
      lookupURL.split("/")[1].toLowerCase() !== DEFAULT_LOCALE.toLowerCase() &&
      !CONTENT_TRANSLATED_ROOT
    ) {
      // Such a common mistake. You try to view a URL that is not en-US but
      // you forgot to set CONTENT_TRANSLATED_ROOT.
      console.warn(
        `URL is for locale '${
          lookupURL.split("/")[1]
        }' but CONTENT_TRANSLATED_ROOT is not set. URL will 404.`
      );
    }
  } catch (error) {
    console.error(`Error in buildDocumentFromURL(${lookupURL})`, error);
    return res.status(500).send(error.toString());
  } finally {
    console.timeEnd(`buildDocumentFromURL(${lookupURL})`);
  }

  if (!document) {
    const redirectURL = Redirect.resolve(lookupURL);
    if (redirectURL !== lookupURL) {
      return res.redirect(301, redirectURL + extraSuffix);
    }
    return res
      .status(404)
      .sendFile(path.join(STATIC_ROOT, "en-us", "_spas", "404.html"));
  }

  if (isDocument) {
    res.json({ doc: document });
  } else if (isMetadata) {
    const docString = JSON.stringify({ doc: document });

    const hash = crypto.createHash("sha256").update(docString).digest("hex");
    const { body: _, toc: __, sidebarHTML: ___, ...builtMetadata } = document;
    builtMetadata.hash = hash;

    res.json(builtMetadata);
  } else if (isJSONRequest) {
    // TODO: what's this for?
    res.json({ doc: document });
  } else {
    res.header("Content-Security-Policy", CSP_VALUE);
    res.send(renderHTML(lookupURL, { doc: document }));
  }
});

if (!fs.existsSync(path.resolve(CONTENT_ROOT))) {
  throw new Error(`${path.resolve(CONTENT_ROOT)} does not exist!`);
}

console.log(
  `CONTENT_ROOT: ${chalk.bold(CONTENT_ROOT)}`,
  path.resolve(CONTENT_ROOT) !== CONTENT_ROOT
    ? chalk.grey(`(absolute path: ${path.resolve(CONTENT_ROOT)})`)
    : ""
);

const PORT = parseInt(process.env.SERVER_PORT || "5042");
app.listen(PORT, () => {
  console.log(`Listening on port ${PORT}`);
  if (process.env.EDITOR) {
    console.log(`Your EDITOR is set to: ${chalk.bold(process.env.EDITOR)}`);
  } else {
    console.warn(
      chalk.yellow(
        "Warning! You have not set an EDITOR environment variable. " +
          'Using the "Open in your editor" button will probably fail.'
      )
    );
  }
});
